design/cluster-update.md

```diff
@@ -0,0 +1,209 @@
+* User updates the Cluster CRD to change the device filter or list to now be `sdb`.
+* The OSD pod restarts and when it comes back up it sees that the previously configured `sda` is no longer in the filter.
+* The pod checks the device UUID of `sda` in its saved config and compares that to the device UUID of the current `sda`.
+* The two match, so the pod knows it's a safe (user intended) operation to remove the OSD from `sda`.
+
+### Orchestration
+When the operator receives an event that the Cluster CRD has been updated, it will need to perform some orchestration in order to bring actual state of the cluster in agreement with the desired state.
+For example, if `monCount` were to be updated, the operator would need to add or remove a single monitor at a time, ensuring that quorum is restored before moving onto the next monitor.
+Updates to the storage spec for the cluster require even more careful consideration and management by the operator, which will be discussed in this section.
+
+First and foremost, changes to the cluster state should not be carried out when the cluster is not in a healthy state.
+The operator should wait until cluster health is restored until any orchestration is carried out.
+
+It is important to remember that a single OSD pod can contain multiple OSD processes and that the operator itself does not have detailed knowledge of the storage resources of each node.
+More specifically, the devices that can be used for storage (e.g., match `deviceFilter`) is not known until the OSD pod has been started on a given node.
+
+As mentioned previously, it is recommended to make storage changes to the cluster one OSD at a time.
+Therefore, the operator and the OSD pods will need to coordinate their efforts in order to adhere to this guidance.
+When a cluster update event is received by the operator, it will work on a node by node basis, ensuring all storage updates are completed by the OSD pod for that node before moving to the next.
+
+When an OSD pod starts up and has completed its device discovery, it will need to perform a diff of the desired storage against the actual storage that is currently included in the cluster.
+This diff will determine the set of OSD instances that need to be removed or added within the pod.
+Fortunately, the OSD pod start up is already idempotent and already handles new storage additions, so the remaining work will be the following:
+* Safely removing existing OSDs from the cluster
+* Waiting for data migration to complete and all placement groups to become clean
+* Signaling to the operator that the pod has completed its storage updates
+
+We should consider an implementation that allows the OSD pod to refresh it's set of OSDs without restarting the entire pod, but since the replication controller's pod template spec needs to be updated by the operator in order to convey this information to the pod, we may need to live with restarting the pod either way.
```

Comment:
Comment: I don't see it as a problem that the pod is restarted. The user should know that when changing something it can "interrupt" the storage (depending on which failure domain e.g. host used and how many data replicas).